### PR TITLE
Add support for middleware to the object service

### PR DESCRIPTION
**Pull request by djmitche**
Github Bug/Issue: Refs #3951 (but doesn't fix it, as it doesn't actually add CDN support)

### CODE REVIEW

**Comment by djmitche** [comment]
Erg, I keep missing expiration_test.js :)

**Comment by djmitche** [comment]
^^ ignore that

**Review by jwhitlock — APPROVED** [review]
Thanks @djmitche. This feels like a shadow of the solution, but I imagine it will become clearer when the CDN implementation is added in the next PR.

In Django, middleware are conceptually wrappers around the application, adding layers like an onion that requests and responses go through. They can process and modify requests on the way in, and then can process and modify requests on the way out. Some are primarily request middlewares, for example checking authentication credentials and exiting early. Others are primarily response middleware, for example adding CSP headers. And some act as both request and response handlers, returning `304 Not Modified` responses based on request cache headers, and setting the cache headers on responses. One quirk of this is that middleware are applied in forward order on requests, and reverse order on responses, so they act like layers.

My concept of a CDN middleware would be to get the "raw" URL from the backend response, and then switch out the domain name when processing the response. I'm guessing instead the CDN middleware needs to do this with the request, before the backend can process it? I'm not sure how that logic will work.

But, that's the next PR. This one looks good, and in theory a middleware response path could be added if needed.

File: services/object/README.md [hunk 1]
```diff
@@ -11,6 +11,14 @@ Each object is assigned a backend when it is created.
 
 The base class for backends, in `src/backends/base.js`, is authoritative for the methods and behaviors that a backend must display.
 
+## Download Middleware
+
+HTTP server developers love the word "middleware"!
```
[the remark this answers is not in the document]
Comment (djmitche) [reply services/object/README.md:16]: Haha, it was intended as an apology for my own overusing the word!  But you're right, it adds nothing.

File: services/object/src/middleware/index.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+      if (!await mw.downloadObjectRequest(req, res, object, method, params)) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  /**
+   * Similar to downloadObjectRequest, but for the simple-downloda API.
```
[the remark this answers is not in the document]
Comment (djmitche) [reply services/object/src/middleware/index.js:50]: I should put that in my vim aliases..

**Comment by djmitche** [comment]
Thanks for the info about middleware in Django.

Yes, I expect CDN to intercept the request, rather than the response, and just translate the object.name into a CDN URL.

You're right that for more complex purposes, we might need to handle responses as well, and allow their modification mid-stream.  As you say, we can add that later -- this is an API that's entirely internal to the service, so we can change it easily.

**Pull request merged**